src/include/miopen/reduce_common.hpp

```diff
@@ -0,0 +1,311 @@
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ *
+ *******************************************************************************/
+#ifndef GUARD_MIOPEN_REDUCE_COMMON_HPP
+#define GUARD_MIOPEN_REDUCE_COMMON_HPP
+
+#include <half.hpp>
+#include <limits>
+#include <cmath>
+#include <miopen/miopen.h>
+#include <miopen/float_equal.hpp>
+
+#include "bfloat16.hpp"
+
+namespace reduce {
+
+enum ReductionMethod_t
+{
+    Reduce_DirectThreadWise = 1,
+    Reduce_DirectWarpWise   = 2,
+    Reduce_BlockWise        = 3,
+    Reduce_MultiBlock       = 4
+};
+
+// data type conversion
+template <typename T>
+struct type_convert
+{
+    template <typename X>
+    T operator()(X x) const
+    {
+        return static_cast<T>(x);
+    }
+};
+
+template <>
+template <>
+inline float type_convert<float>::operator()<half_float::half>(half_float::half x) const
+{
+    return half_float::half_cast<float>(x);
+};
+
+template <>
+template <>
+inline half_float::half type_convert<half_float::half>::operator()<float>(float x) const
+{
+    return half_float::half_cast<half_float::half>(x);
+};
+
+template <>
+template <>
+inline float type_convert<float>::operator()<bfloat16>(bfloat16 x) const
+{
+    return float(x);
+};
+
+template <>
+template <>
+inline bfloat16 type_convert<bfloat16>::operator()<float>(float x) const
+{
+    return bfloat16(x);
+};
+
+template <typename compType>
+static inline std::function<void(compType&, compType)> ReduceOpFn(miopenReduceTensorOp_t op_)
+{
+    switch(op_)
+    {
+    case MIOPEN_REDUCE_TENSOR_ADD: return ([&](compType& a_, compType b_) { a_ = a_ + b_; });
+
+    case MIOPEN_REDUCE_TENSOR_MUL: return ([&](compType& a_, compType b_) { a_ = a_ * b_; });
+
+    case MIOPEN_REDUCE_TENSOR_MIN:
+        return ([&](compType& a_, compType b_) {
+            if(a_ > b_)
+                a_ = b_;
+        });
+
+    case MIOPEN_REDUCE_TENSOR_MAX:
+        return ([&](compType& a_, compType b_) {
+            if(a_ < b_)
+                a_ = b_;
+        });
+    }
+
+    return (std::function<void(compType&, compType)>{});
+};
+
+template <typename compType>
+static inline std::function<void(compType&, compType, bool& changed)>
+ReduceOpFn2(miopenReduceTensorOp_t op_)
+{
+    switch(op_)
+    {
+    case MIOPEN_REDUCE_TENSOR_MIN:
+        return ([&](compType& a_, compType b_, bool& changed) {
+            if(a_ > b_)
+            {
+                a_      = b_;
+                changed = true;
+            }
+            else
+                changed = false;
```

Comment:
What is this variable used for?

Reply:
It is for the case when the indices of the reduced values need be recorded.  "Changed == true" indicates the index need also be updated to the accuIndex.